python/revng/tupletree.py

```diff
@@ -128,6 +128,24 @@ def __setattr__(self, key, value):
         super().__setattr__(key, value)
 
 
+class UpcastableStructBase(StructBase):
+    _children: Dict[str, Type] = {}
+
+    @classmethod
+    def from_dict(cls, **kwargs):
+        if "Kind" in kwargs:
+            child_cls = cls._children.get(kwargs["Kind"])
+            if child_cls:
+                if cls != child_cls:
+                    return child_cls.from_dict(**kwargs)
+                else:
+                    return super().from_dict(**kwargs)
+            else:
+                raise ValueError(f"No class found to deserialize {kwargs['Kind']}")
+        else:
+            raise ValueError("Upcastable types must have a Kind field")
```

Comment:
Do not nest unnecessarily: go from
```python
if a:
  ...
else:
  raise
```
to:
```python
if not a:
  raise
...
```

Reply:
Fixed